build_tools/cmake/build_docs.sh

```diff
@@ -61,63 +61,59 @@ PYTHON_BIN=`which python3` scripts/update_e2e_coverage.py ${BUILD_DIR}
-  llvm-ir
-  vmla
-  vulkan-spirv
-)
+build_one_example() {
+  example=$1
+  target=$2
 
-mkdir -p ${BUILD_DIR}/doc/ir_examples
-for example in "${ir_examples[@]}"; do
-  for target in "${compiler_targets[@]}"; do
-    example_name=$(basename ${example%.mlir})
-    filepath=${BUILD_DIR}/doc/ir_examples/${example_name}-${target}.md
-    title="${example_name} on ${target}"
+  example_name=$(basename ${example%.mlir})
+  filepath=${BUILD_DIR}/doc/ir_examples/${example_name}-${target}.md
+  title="${example_name} on ${target}"
 
-    echo -e "# ${title}\n" > ${filepath}
+  echo -e "# ${title}\n" > ${filepath}
 
-    preamble="
+  preamble="
 This is an IR conversion example auto-generated by running \`iree-opt\`
 over [${example_name}](https://github.com/google/iree/tree/main/${example}).
 
 ### The command
 "
-    echo -e "${preamble}" >> ${filepath}
-    echo -e -n '```shell' >> ${filepath}
+  echo -e "${preamble}" >> ${filepath}
+  echo -e -n '```shell' >> ${filepath}
 
-    cmd="
+  cmd="
 iree-opt -iree-transformation-pipeline \\
   -iree-hal-target-backends=${target} \\
   --print-ir-after-all \\
   -mlir-disable-threading \\
   -mlir-elide-elementsattrs-if-larger=8 \\
   ${example}
 "
-    echo -e "${cmd}" >> ${filepath}
-    echo -e '```' >> ${filepath}
-
-    echo -e "### Input IR\n" >> ${filepath}
-    echo -e '```' >> ${filepath}
-    cat ${example} >> ${filepath}
-
-    tmpfile=$(mktemp)
-    ${BUILD_DIR}/iree/tools/iree-opt \
-      -iree-transformation-pipeline \
-      -iree-hal-target-backends=${target} \
-      -print-ir-after-all \
-      -mlir-disable-threading \
-      -mlir-elide-elementsattrs-if-larger=8 \
-      ${example} 1>/dev/null 2>${tmpfile}
-    # Turn pass comment into headers
-    sed 's!^// \*\*\* IR Dump After \(.*\) \*\*\*$!```\n### IR Dump After \1\n```!' \
-      ${tmpfile} >> ${filepath}
-    # Remove extra empty lines
-    sed -i '/^$/N;/^\n$/D' ${filepath}
-    echo -e '```' >> ${filepath}
-  done
-done
+  echo -e "${cmd}" >> ${filepath}
+  echo -e '```' >> ${filepath}
+
+  echo -e "### Input IR\n" >> ${filepath}
+  echo -e '```' >> ${filepath}
+  cat ${example} >> ${filepath}
+
+  tmpfile=$(mktemp)
+  ${BUILD_DIR}/iree/tools/iree-opt \
+    -iree-transformation-pipeline \
+    -iree-hal-target-backends=${target} \
+    -print-ir-after-all \
+    -mlir-disable-threading \
+    -mlir-elide-elementsattrs-if-larger=8 \
+    ${example} 1>/dev/null 2>${tmpfile}
+  # Turn pass comment into headers
+  sed 's!^// \*\*\* IR Dump After \(.*\) \*\*\*$!```\n### IR Dump After \1\n```!' \
+    ${tmpfile} >> ${filepath}
+  # Remove extra empty lines
+  sed -i '/^$/N;/^\n$/D' ${filepath}
+  echo -e '```' >> ${filepath}
+}
+
+mkdir -p ${BUILD_DIR}/doc/ir_examples
+
+build_one_example iree/samples/ops/mhlo-dot.mlir llvm-ir
+build_one_example iree/samples/ops/mhlo-dot.mlir vmla
+build_one_example iree/samples/ops/mhlo-dot.mlir vulkan-spirv
+build_one_example iree/samples/ops/dynamic-mhlo-dot.mlir vmla
```

Comment:
Nice, this reads more clearly than the previous setup.

If the `build_one_example` function itself gets more complex, we should consider moving it into a separate (maybe Python?) file.

Reply:
Agreed. I think we can use MLIR hooks to capture a more structured trace. Maybe we can generate sliders or something more dynamic showing the IR evolution over time or something.